### PR TITLE
Update `replace!` function for multiple `Tensor` replacements and fix `Symbol` check

### PR DESCRIPTION
**Pull request by jofrevalles**
### Summary
This PR includes several improvements to the `replace!` function in the `TensorNetwork` module:

1. Fixed the `replace!` function for `Symbol` pairs to correctly check if target symbols are not already present in the network. The line `!isdisjoint(values(old_new), labels(tn)) && throw(...)` was updated to `!isdisjoint(last.(old_new), labels(tn)) && throw(...)`.
2. Updated the `replace!` function for `Tensor` pairs to now accept multiple replacements instead of just one. This allows for more efficient and convenient replacement of multiple tensors within a network.
3. Added new tests to ensure the correctness of the updated `replace!` function for both `Symbol` and `Tensor` pairs. The new tests check the previous error in the `replace!` function for symbols and the functionality of multiple tensor replacements.

### Example
Example of multiple replacements of `Tensor`s:
```julia
julia> using Tenet

julia> using Test

julia> t_ij = Tensor(rand(2, 2), (:i, :j))

julia> t_ik = Tensor(rand(2, 2), (:i, :k))

julia> t_ilm = Tensor(rand(2, 2, 2), (:i, :l, :m))

julia> t_lm = Tensor(rand(2, 2), (:l, :m))

julia> tn = TensorNetwork([t_ij, t_ik, t_ilm, t_lm])
TensorNetwork{Arbitrary}(#tensors=4, #inds=5)

julia> new_tensor1 = Tensor(rand(2, 2), (:i, :k))
2×2 Tensor{Float64, 2, Matrix{Float64}}:
 0.225687  0.85958
 0.104862  0.00510741

julia> new_tensor2 = Tensor(ones(2, 2), (:i, :k))
2×2 Tensor{Float64, 2, Matrix{Float64}}:
 1.0  1.0
 1.0  1.0

julia> replace!(tn, t_ik=>new_tensor1, new_tensor1=>new_tensor2)
TensorNetwork{Arbitrary}(#tensors=4, #inds=5)

julia> tensors(tn, 2)
2×2 Tensor{Float64, 2, Matrix{Float64}}:
 1.0  1.0
 1.0  1.0
```

### CODE REVIEW

**Comment by mofeing** [comment]
This seems like a rewrite of something that *should* already work (maybe with some bug fixes) and the refactor seems much more verbose.

> 1. Fixed the `replace!` function for `Symbol` pairs to correctly check if target symbols are not already present in the network. The line `!isdisjoint(values(old_new), labels(tn)) && throw(...)` was updated to `!isdisjoint(last.(old_new), labels(tn)) && throw(...)`.

Okay, super. My bad for using `values` instead of `last`.

> 2. Updated the `replace!` function for `Tensor` pairs to now accept multiple replacements instead of just one. This allows for more efficient and convenient replacement of multiple tensors within a network.

mmm this was already working. Was there any bug?

> 3. Added new tests to ensure the correctness of the updated `replace!` function for both `Symbol` and `Tensor` pairs. The new tests check the previous error in the `replace!` function for symbols and the functionality of multiple tensor replacements.

Nice.

In general, I don't understand the reasons behind the refactor. Could explain yourself a little bit?

File: test/TensorNetwork_test.jl [hunk 1]
```diff
@@ -263,6 +265,20 @@
...
             @test new_tensor === tensors(tn, 1)
             @test B === tensors(tn, 2)
+
+            # Test chain of replacements
+            A = Tensor(zeros(2, 2), (:i, :j))
+            B = Tensor(zeros(2, 2), (:j, :k))
+            C = Tensor(zeros(2, 2), (:k, :l))
+            tn = TensorNetwork([A, B, C])
+
+            @test_throws ArgumentError replace!(tn, A => B, B => C, C => A)
```
Comment (mofeing) [inline test/TensorNetwork_test.jl:275]: This should work.

File: test/TensorNetwork_test.jl [hunk 2]
```diff
@@ -263,6 +265,20 @@
...
+
+            new_tensor = Tensor(rand(2, 2), (:i, :j))
+            new_tensor2 = Tensor(ones(2, 2), (:i, :j))
+
+            replace!(tn, A => new_tensor, new_tensor => new_tensor2)
+            @test new_tensor2 === tensors(tn, 1)
```
Comment (mofeing) [inline test/TensorNetwork_test.jl:281]: And this should not work.

File: src/TensorNetwork.jl [hunk 3]
```diff
@@ -45,25 +45,33 @@ function TensorNetwork{A}(tensors; meta...) where {A}
...
-function Base.replace!(tn::TensorNetwork, pair::Pair{<:Tensor, <:Tensor})
-    old_tensor, new_tensor = pair
+Base.replace(tn::TensorNetwork, old_new::Pair...) = replace!(copy(tn), old_new...)
 
-    # check if old and new tensors are compatible
-    if !issetequal(labels(new_tensor), labels(old_tensor))
-        throw(ArgumentError("New tensor labels do not match the existing tensor labels"))
+function Base.replace!(tn::TensorNetwork, old_new::Pair{<:Tensor, <:Tensor}...)
+    # Check if new tensors are already present in the network
+    new_tensors = [new_tensor for (_, new_tensor) in old_new]
```
Comment (mofeing) [inline src/TensorNetwork.jl:52]:
```suggestion
    new_tensors = last.(old_new)
```

File: src/TensorNetwork.jl [hunk 4]
```diff
@@ -45,25 +45,33 @@ function TensorNetwork{A}(tensors; meta...) where {A}
...
+            link_index = findfirst(x -> x === old_tensor, index_obj.links)
+            index_obj.links[link_index] = new_tensor
+        end
 
-    # replace existing `Tensor` with new `Tensor`
-    index = findfirst(x -> x === old_tensor, tn.tensors)
-    tn.tensors[index] = new_tensor
+        # replace existing `Tensor` with new `Tensor`
+        index = findfirst(x -> x === old_tensor, tn.tensors)
+        tn.tensors[index] = new_tensor
```
Comment (mofeing) [inline src/TensorNetwork.jl:73]: Use `pop!` and `push!`.

**Review by mofeing — CHANGES_REQUESTED** [review]

File: test/TensorNetwork_test.jl (hunk 1 above)
Comment (jofrevalles) [reply test/TensorNetwork_test.jl:275]: Why? We talked that the `replace` function just replaces tensors with the same `labels`.

File: test/TensorNetwork_test.jl (hunk 2 above)
Comment (jofrevalles) [reply test/TensorNetwork_test.jl:281]: I don't get why it shouldn't.

File: src/TensorNetwork.jl (hunk 4 above)
Comment (jofrevalles) [reply src/TensorNetwork.jl]: Why? `pop!` removes the last element in the array.

File: src/TensorNetwork.jl [hunk 5]
```diff
@@ -223,10 +200,37 @@ end
...
+Base.replace(tn::TensorNetwork, old_new::Pair...) = replace!(copy(tn), old_new...)
+
+function Base.replace!(tn::TensorNetwork, old_new::Pair{<:Tensor,<:Tensor}...)
+    # check if new tensors are already present in the network
+    new_tensors = last.(old_new)
+    isdisjoint(new_tensors, tn.tensors) ||
+        throw(ArgumentError("New tensors must not be already present in the network"))
+
+    # check if old and new tensors are compatible
+    all(pair -> issetequal(labels(pair.first), labels(pair.second)), old_new) ||
```
Comment (mofeing) [inline src/TensorNetwork.jl:212]:
```suggestion
    all(pair -> issetequal(map(labels, pair)...), old_new) ||
```

File: src/TensorNetwork.jl [hunk 6]
```diff
@@ -223,10 +200,37 @@ end
 
 Base.delete!(tn::TensorNetwork, x) = (_ = pop!(tn, x); tn)
 
-Base.replace(tn::TensorNetwork, old_new::Pair{Symbol,Symbol}...) = replace!(copy(tn), old_new...)
+Base.replace(tn::TensorNetwork, old_new::Pair...) = replace!(copy(tn), old_new...)
+
+function Base.replace!(tn::TensorNetwork, old_new::Pair{<:Tensor,<:Tensor}...)
+    # check if new tensors are already present in the network
+    new_tensors = last.(old_new)
+    isdisjoint(new_tensors, tn.tensors) ||
```
Comment (mofeing) [inline src/TensorNetwork.jl:208]:
```suggestion
    isdisjoint(last.(old_new), tn.tensors) ||
```

**Comment by jofrevalles** [comment]
@mofeing I think that now the branch is ready to merge.

File: test/TensorNetwork_test.jl (hunk 1 above)
Comment (mofeing) [reply test/TensorNetwork_test.jl:275]: You're right.

File: src/TensorNetwork.jl (hunk 4 above)
Comment (mofeing) [reply src/TensorNetwork.jl]: You're right. Maybe we have to revisit this idea when #55 is merged.

**Pull request merged**